src/test/java/util/CacheTest.java

```diff
@@ -0,0 +1,283 @@
+		assertThat(loaderCallCount.get()).isEqualTo(2);
+
+		// Invalidate one entry
+		cache.invalidate("key1");
+
+		// Next get should invoke loader again for invalidated key
+		assertThat(cache.get("key1")).isEqualTo("value-key1");
+		assertThat(loaderCallCount.get()).isEqualTo(3);
+
+		// Non-invalidated key should still be cached
+		assertThat(cache.get("key2")).isEqualTo("value-key2");
+		assertThat(loaderCallCount.get()).isEqualTo(3);
+	}
+
+	@Test
+	public void testInvalidateAll() {
+		Cache<String, String> cache = new Cache.Builder<String, String>()
+				.build(this::increment);
+
+
+		// Add entries
+		assertThat(cache.get("key1")).isEqualTo("value-key1");
+		assertThat(cache.get("key2")).isEqualTo("value-key2");
+		assertThat(loaderCallCount.get()).isEqualTo(2);
+
+		// Invalidate all entries
+		cache.invalidateAll();
+		assertThat(cache.size()).isEqualTo(0);
+
+		// Next gets should invoke loader again
+		assertThat(cache.get("key1")).isEqualTo("value-key1");
+		assertThat(cache.get("key2")).isEqualTo("value-key2");
+		assertThat(loaderCallCount.get()).isEqualTo(4);
+	}
+
+	@Test
+	public void testConcurrentAccess() throws InterruptedException {
+		Cache<String, String> cache = new Cache.Builder<String, String>()
+				.build(key -> {
+					// Add some delay to increase chance of concurrent issues
+					try {
+						Thread.sleep(10);
+					} catch (InterruptedException e) {
+						// Ignore
+					}
+					loaderCallCount.incrementAndGet();
+					return "value-" + key;
+				});
+
+		// Create multiple threads that access the same key
+		int threadCount = 10;
+		Thread[] threads = new Thread[threadCount];
+
+		for (int i = 0; i < threadCount; i++) {
+			threads[i] = new Thread(() -> {
+				assertThat(cache.get("shared")).isEqualTo("value-shared");
+			});
+		}
+
+		// Start all threads
+		for (Thread thread : threads) {
+			thread.start();
+		}
+
+		// Wait for all threads to complete
+		for (Thread thread : threads) {
+			thread.join();
+		}
+
+		// Should only call the loader once despite multiple threads
+		assertThat(loaderCallCount.get()).isEqualTo(1);
+	}
+
+	@Test
+	public void testBothExpirationStrategies() throws InterruptedException {
+		Cache<String, String> cache = new Cache.Builder<String, String>()
+				.expireAfterWrite(Duration.ofMillis(200))
+				.expireAfterAccess(Duration.ofMillis(100))
+				.build(this::increment);
+
+
+		// First call should invoke loader
+		assertThat(cache.get("key1")).isEqualTo("value-key1");
+		assertThat(loaderCallCount.get()).isEqualTo(1);
+
+		assertThat(cache.get("key2")).isEqualTo("value-key2");
+		assertThat(loaderCallCount.get()).isEqualTo(2);
+
+		// Access within access expiration time
+		Thread.sleep(50); // running total: 50ms
+		assertThat(cache.get("key1")).isEqualTo("value-key1");
+		assertThat(loaderCallCount.get()).isEqualTo(2);
+
+		// Wait for access expiration but not write expiration
+		Thread.sleep(100); // running total: 150ms
```

Comment:
I've merged but the tests have failed on the main branch. I think this 100 needs to be bumped a little to make this test more consistent, as it had passed on this branch before merging
This may be because the access expiry is 100ms and the wait is only 100 ms so may be just within the threshold.